solr/core/src/java/org/apache/solr/managed/ResourceManager.java

```diff
@@ -0,0 +1,305 @@
+
+  /**
+   * Modify parameters of an existing pool.
+   * @param name existing pool name
+   * @param params new parameter values.
+   * @throws Exception when an invalid value or unsupported parameter is requested, or the parameter
+   * value cannot be changed after creation
+   */
+  public abstract void setPoolParams(String name, Map<String, Object> params) throws Exception;
+
+  /**
+   * Remove pool. This also stops the management of resources registered with that pool.
+   * @param name existing pool name
+   */
+  public abstract void removePool(String name) throws Exception;
+
+  /**
+   * Add managed components to a pool.
+   * @param pool existing pool name.
+   * @param managedComponents components to add
+   */
+  public void registerComponents(String pool, Collection<ManagedComponent> managedComponents) {
+    ensureActive();
+    for (ManagedComponent managedComponent : managedComponents) {
+      registerComponent(pool, managedComponent);
+    }
+  }
+
+  /**
+   * Add a managed component to a pool.
+   * @param pool existing pool name.
+   * @param managedComponent managed component. The component must support the management type
+   *                        used in the selected pool. The component must not be already managed by
+   *                        another pool of the same type.
+   */
+  public abstract void registerComponent(String pool, ManagedComponent managedComponent);
+
+  /**
+   * Remove a managed component from a pool.
+   * @param pool existing pool name.
+   * @param componentId component id to remove
+   * @return true if a component was actually registered and has been removed
+   */
+  public boolean unregisterComponent(String pool, ManagedComponentId componentId) {
+    return unregisterComponent(pool, componentId.toString());
+  }
+
+  /**
+   * Unregister component from all pools.
+   * @param componentId component id
+   * @return true if a component was actually registered and removed from at least one pool.
+   */
+  public boolean unregisterComponent(String componentId) {
+    boolean removed = false;
+    for (String pool : listPools()) {
+      removed = removed || unregisterComponent(pool, componentId);
+    }
+    return removed;
+  }
+
+  /**
+   * Remove a managed component from a pool.
+   * @param pool existing pool name.
+   * @param componentId component id to remove
+   * @return true if a component was actually registered and has been removed
+   */
+  public abstract boolean unregisterComponent(String pool, String componentId);
+
+  protected void ensureActive() {
+    if (isClosed) {
+      throw new IllegalStateException("Already closed.");
+    }
+    if (!initialized) {
+      throw new IllegalStateException("Not initialized.");
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    synchronized (this) {
+      isClosed = true;
+      SolrMetricProducer.super.close();
+    }
```

Comment:
The usage of isClosed and initialized in this class is not thread safe. Different threads read and set these values so the latest values are not guaranteed to be visible. Both can be made AtomicBoolean instances and then the synchronized(this) block here is not needed at all.

Reply:
Fixed.